vllm/model_executor/models/llama.py

```diff
@@ -122,27 +156,51 @@ def __init__(
         self.rope_theta = rope_theta
         self.max_position_embeddings = max_position_embeddings
 
-        self.qkv_proj = QKVParallelLinear(
-            hidden_size=hidden_size,
-            head_size=self.head_dim,
-            total_num_heads=self.total_num_heads,
-            total_num_kv_heads=self.total_num_kv_heads,
-            bias=bias,
-            quant_config=quant_config,
-        )
+        self.merge_weight = getattr(quant_config, "merge_weight", True)
+        if self.merge_weight:
+            self.qkv_proj = QKVParallelLinear(
+                hidden_size=hidden_size,
+                head_size=self.head_dim,
+                total_num_heads=self.total_num_heads,
+                total_num_kv_heads=self.total_num_kv_heads,
+                bias=bias,
+                quant_config=quant_config,
+            )
+        else:
+            self.q_proj = ColumnParallelLinear(
+                hidden_size,
+                self.total_num_heads * self.head_dim,
+                bias=bias,
+                quant_config=quant_config,
+            )
+            self.k_proj = ColumnParallelLinear(
+                hidden_size,
+                self.total_num_kv_heads * self.head_dim,
+                bias=bias,
+                quant_config=quant_config,
+            )
+            self.v_proj = ColumnParallelLinear(
+                hidden_size,
+                self.total_num_kv_heads * self.head_dim,
+                bias=bias,
+                quant_config=quant_config,
+            )
+
         self.o_proj = RowParallelLinear(
             input_size=self.total_num_heads * self.head_dim,
             output_size=hidden_size,
             bias=bias,
             quant_config=quant_config,
         )
 
+        is_neox_style = getattr(quant_config, "is_neox_style", True)
         self.rotary_emb = get_rope(
             self.head_dim,
             rotary_dim=self.head_dim,
             max_position=max_position_embeddings,
             base=rope_theta,
             rope_scaling=rope_scaling,
+            is_neox_style=is_neox_style,
```

Comment:
Why are we setting neox_style for llama? I don't see why the quant config controls this

Reply:
Hmmm, there is an issue about the rotary_emb difference between official llama and hf llama implement/checkpoints (also see https://github.com/huggingface/transformers/issues/25199#issuecomment-1687720247). 

The hf implement permutes the weights and uses neox_style to optimize the rotary_emb. However, llama implement in llama.cpp uses the original rotary_emb and doesn't use neox_style one's [(llama.cpp#L19344-L19401)](https://github.com/ggerganov/llama.cpp/blob/master/src/llama.cpp#L19344-L19401), so we need to set `is_neox_style=False` to make rotary_emb work correct.

I agree that we should remove `is_neox_style` property from `GGUFConfig`. Would this code below be a good choice?
```python
is_neox_style = False if quant_config.get_name=="gguf" else True
```